src/main/scala/uk/gov/nationalarchives/tdr/api/service/FileMetadataService.scala

```diff
@@ -4,33 +4,40 @@ import java.sql.{SQLException, Timestamp}
 import java.util.UUID
 
 import uk.gov.nationalarchives.Tables.{FilemetadataRow, FilepropertyRow}
-import uk.gov.nationalarchives.tdr.api.db.repository.{FileMetadataRepository, FilePropertyRepository}
+import uk.gov.nationalarchives.tdr.api.db.repository.{ClientFileMetadataRepository, FileMetadataRepository, FilePropertyRepository, FileRepository}
 import uk.gov.nationalarchives.tdr.api.graphql.DataExceptions.InputDataException
 import uk.gov.nationalarchives.tdr.api.graphql.fields.FileMetadataFields.{AddFileMetadataInput, FileMetadata}
 
 import scala.concurrent.{ExecutionContext, Future}
 
 class FileMetadataService(fileMetadataRepository: FileMetadataRepository, filePropertyRepository: FilePropertyRepository,
+                          clientFileMetadataService: ClientFileMetadataService,
                           timeSource: TimeSource, uuidSource: UUIDSource)(implicit val ec: ExecutionContext) {
 
   def addFileMetadata(addFileMetadataInput: AddFileMetadataInput, userId: Option[UUID]): Future[FileMetadata] = {
-    getFileProperty(addFileMetadataInput.filePropertyName).flatMap {
-      case Some(property) =>
-        val row = FilemetadataRow(uuidSource.uuid,
-          addFileMetadataInput.fileId,
-          property.propertyid,
-          addFileMetadataInput.value,
-          Timestamp.from(timeSource.now),
-          userId.get)
-
-        fileMetadataRepository.addFileMetadata(row).map(r => FileMetadata(property.name.get, r.fileid, r.value)).recover {
-          case e: SQLException => throw InputDataException(e.getLocalizedMessage, Some(e))
+    def row(property: FilepropertyRow) = FilemetadataRow(uuidSource.uuid,
+      addFileMetadataInput.fileId,
+      property.propertyid,
+      addFileMetadataInput.value,
+      Timestamp.from(timeSource.now),
+      userId.get)
+
+    addFileMetadataInput.filePropertyName match {
+      case "SHA256ServerSideChecksum" =>
+        (for {
+          fileProperty <- getFileProperty(addFileMetadataInput.filePropertyName)
+          cfm <- clientFileMetadataService.getClientFileMetadata(addFileMetadataInput.fileId) if fileProperty.isDefined
+          row <- fileMetadataRepository.addChecksumMetadata(row(fileProperty.get), cfm.checksum.map(_ == addFileMetadataInput.value))
+        } yield FileMetadata(fileProperty.flatMap(_.name).get, row.fileid, row.value)) recover {
+          case e: Throwable => throw InputDataException(e.getLocalizedMessage, Some(e))
         }
-      case None => throw InputDataException(s"The property does not exist", Option.empty)
+      case _ => Future.failed(InputDataException("We are only expecting checksum updates for now"))
```

Comment:
Can you stick the actual `filePropertyName` in the error message so we can work out what happened if we ever hit this error?

Reply:
Yep, makes sense, I've added it in.

Reply:
That's useful, thanks, but I meant adding `addFileMetadataInput.filePropertyName`. An error like "Unexpected property 'MD5Hash'" is easier to debug than just "Unexpected property" because you can start asking "Wait, what's trying to save MD5 hashes?" 😄

Reply:
That's what I've done I think https://github.com/nationalarchives/tdr-consignment-api/pull/76/files#diff-a2f15bd61b5d5ae2b2905333419fa072R36
It adds the property name in. Or is it the wording of the error you mean?

Reply:
Ahh, I think I was somehow looking at an older version of the code. Sorry!